### PR TITLE
Folders: Access parent folder without view permission POC

### LINKED ISSUE #71425: Accessing a folder without permissions to the parent
**Problem:**

Currently, the cascading permissions for subfolders work well in most scenarios. However, there are cases where a team has access to a subfolder and its contents, but not the parent folder.

In such cases, the subfolder does not appear in the Browse view. The team members need to search for the folder by name, which is not an efficient process. Moreover, if they are not aware of the folder's existence or its name, they might not be able to access it at all.

**Expected Behavior:**

Teams should be able to access their folders in a more straightforward manner, directly from the Browse view, even if they do not have access to the parent folder.

**Solution:**

The subfolder to which the user has access will be displayed as if it were a root-level folder. This is because the user doesn't have access to the parent folder, so from their perspective, the subfolder is the highest-level folder they can see.

### PR DESCRIPTION
**Pull request by alexanderzobnin**
**What is this feature?**

This is POC for #71425

It allow to display full subtree of the folders that user has access to, including parent folders without "view" permission.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #71425

**Special notes for your reviewer:**

So idea for this POC is that we can implement a kind of "traverse" permission which allows to access folders that parents for those with view access for the particular user.

nested_folders_permissions_01

The algorithm is pretty straightforward (I haven't checked performance yet, but the fact that we have `GetParents()` method which utilized SQL built-in recurse if possible sounds promising): 
1. Collect list of folders that user has access to (extract from user's permissions).
2. For each folder get list of the parents ( `GetParents()` returns all parents until the root folder). Most performance-demanding part.
3. When evaluating access to the folder, check if current user has view access to folder or it's in the list of parents (user can "traverse") folder.

nested_folders_permissions_02

Idea is similar to Unix execution (`x`) bit for directories, it does not allow to read dir, but traverse it.

From the performance perspective, I need to investigate the possibilities of optimization. Looks like one of the way is to cache list of parent folders for particular user since it's most complicated part.

### CODE REVIEW

File: pkg/services/guardian/accesscontrol_guardian.go [hunk 1]
```diff
@@ -282,6 +300,42 @@ func (a *accessControlFolderGuardian) CanView() (bool, error) {
 	return a.evaluate(accesscontrol.EvalPermission(dashboards.ActionFoldersRead, dashboards.ScopeFoldersProvider.GetResourceScopeUID(a.folder.UID)))
 }
 
+func (a *accessControlFolderGuardian) CanTraverse() (bool, error) {
+	if a.folder == nil {
+		return false, ErrGuardianFolderNotFound.Errorf("failed to check permissions for folder")
+	}
+
+	folderPermissions := a.user.Permissions[a.user.OrgID]["dashboards:read"]
```
Comment (alexanderzobnin) [inline pkg/services/guardian/accesscontrol_guardian.go:308]: This is a part I rely on. Is it correct that `user.Permissions` always holds effective set of permissions user has, including managed permissions, permissions from teams and roles?

File: pkg/services/guardian/accesscontrol_guardian.go [hunk 2]
```diff
@@ -282,6 +300,42 @@ func (a *accessControlFolderGuardian) CanView() (bool, error) {
...
+			folderUids = append(folderUids, folderUid)
+			if folderUid == a.folder.UID {
+				return true, nil
+			}
+		}
+	}
+
+	parentUids := make([]string, 0)
+	for _, uid := range folderUids {
+		parents, err := a.folderService.GetParents(a.ctx, folder.GetParentsQuery{
```
Comment (torkelo) [inline pkg/services/guardian/accesscontrol_guardian.go:321]: Needs to work with a lot folders, ~10 - 50k not sure what should be our target here. thoughts @bergquist ? 

Doing SQL queries in loops is an obvious no no, maybe we can cache this, is folderService.getParents dependent on the calling user? (does it need to be cached for each user? that could get memory intensive).

**Review by torkelo — COMMENTED** [review]
Wonder if there is some way we can precompute some of these nested permission scopes when the permission is created (and when folders move). so when we need to evaluate there is no need for recursive lookups or doing SQL queries in loops

File: pkg/services/guardian/accesscontrol_guardian.go (hunk 2 above)
Comment (alexanderzobnin) [reply pkg/services/guardian/accesscontrol_guardian.go:321]: `GetParents()` does not depend on user, so might be cached, however, huge amount of folders might be a problem.

**Comment by alexanderzobnin** [comment]
@torkelo pre-computing is one of the option I thought about, we need a hook on permissions and folders structure update. This might be quite effective since permissions and folders are not frequently updated.

**Comment by alexanderzobnin** [comment]
Closing in favor of #74273

**Pull request closed without merging**